File: src/main/java/shopping/product/infrastructure/ProductEntity.java
package shopping.product.infrastructure;

import jakarta.persistence.*;
import org.hibernate.annotations.SQLRestriction;
import shopping.common.BaseEntity;
import shopping.product.domain.Product;

import java.math.BigDecimal;

@Entity
@Table(name = "product")
@SQLRestriction("deleted_at IS NULL")
public class ProductEntity extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String name;
    private BigDecimal price;

    @Column(name = "image_url")
    private String imageUrl;

    protected ProductEntity() {}

    private ProductEntity(Long id, String name, BigDecimal price, String imageUrl) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.imageUrl = imageUrl;
    }

    public Product toDomain() {
        return Product.of(id, name, price, imageUrl);
    }

    public static ProductEntity from(Product product) {
        return new ProductEntity(product.getId(), product.getName(), product.getPrice(), product.getImageUrl());
    }

    public void update(Product product) {
        this.name = product.getName();
        this.price = product.getPrice();
        this.imageUrl = product.getImageUrl();
    }
}

File: src/main/kotlin/shopping/storage/db/ProductEntity.kt
package shopping.storage.db

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Table
import shopping.domain.Product

@Entity
@Table(name = "products")
class ProductEntity(
    @Column(nullable = false, length = 15)
    var name: String,
    @Column(nullable = false)
    var price: Int,
    @Column(nullable = false)
    var imageUrl: String,
) : BaseEntity() {
    fun toDomain(): Product =
        Product(
            id = this.id,
            name = this.name,
            price = this.price,
            imageUrl = this.imageUrl,
        )

    fun update(product: Product) {
        this.name = product.name.value
        this.price = product.price.value
        this.imageUrl = product.imageUrl
    }

    companion object {
        fun from(product: Product): ProductEntity =
            ProductEntity(
                name = product.name.value,
                price = product.price.value,
                imageUrl = product.imageUrl,
            )
    }
}
